databuilder/query_model.py

```diff
@@ -1,452 +1,356 @@
-            return node
-
-        operator, value = list(kwargs.items())[0]
-        if operator == "between":
-            # convert a between filter into its two components
-            return self.filter(*args, on_or_after=value[0], on_or_before=value[1])
-
-        if operator in ("equals", "not_equals") and isinstance(
-            value, (Codelist, Column)
-        ):
-            raise TypeError(
-                f"You can only use '{operator}' to filter a column by a single value.\n"
-                f"To filter using a {value.__class__.__name__}, use 'is_in/not_in'."
-            )
-
-        if operator == "is_in" and not isinstance(value, (Codelist, Column)):
-            # convert non-codelist in values to tuple
-            value = tuple(value)
-        assert len(args) == len(kwargs) == 1
-
-        operator = _OPERATOR_MAPPING[operator]
-        return FilteredTable(
-            source=self,
-            column=args[0],
-            operator=operator,
-            value=value,
-            or_null=include_null,
-        )
+class OneRowPerPatientTable(Table):
+    ...
 
-    def earliest(self, *columns):
-        columns = columns or ("date",)
-        return self.first_by(*columns)
-
-    def latest(self, *columns):
-        columns = columns or ("date",)
-        return self.last_by(*columns)
-
-    def first_by(self, *columns):
-        assert columns
-        return Row(source=self, sort_columns=columns, descending=False)
-
-    def last_by(self, *columns):
-        assert columns
-        return Row(source=self, sort_columns=columns, descending=True)
-
-    def date_in_range(
-        self, date, start_column="date_start", end_column="date_end", include_null=True
-    ):
-        """
-        A filter that returns rows for which a date falls between a start and end date (inclusive).
-        Null end date values are included by default
-        """
-        return self.filter(start_column, less_than_or_equals=date).filter(
-            end_column, greater_than_or_equals=date, include_null=include_null
-        )
 
-    def exists(self, column="patient_id"):
-        return self.aggregate("exists", column)
+class ManyRowsPerPatientTable(Table):
+    ...
 
-    def count(self, column="patient_id"):
-        return self.aggregate("count", column)
 
-    def sum(self, column):  # noqa: A003
-        return self.aggregate("sum", column)
+class OneRowPerPatientColumn(Column):
+    ...
 
-    def aggregate(self, function, column):
-        output_column = f"{column}_{function}"
-        row = RowFromAggregate(self, function, column, output_column)
-        return ValueFromAggregate(row, output_column)
 
+class ManyRowsPerPatientColumn(Column):
+    ...
 
-@dataclass(frozen=True)
-class Table(BaseTable):
+
+class SortedTable(ManyRowsPerPatientTable):
+    ...
+
+
+# OPERATIONS
+#
+# Operations indicate the kind of thing they return by subclassing one of the basic
+# types above.
+
+
+class SelectTable(ManyRowsPerPatientTable):
```

Comment:
I'm unsure how to mentally parse `SelectTable`.

Are we selecting a table?  Or `SELECT`ing from a table?  Or something else?

Reply:
We're selecting a table, but I agree it's a bad choice of verb here. Any better suggestions? Off the top of my head ...

`FrameFromTable` — though I quite liked the fact that the public API didn't need to mention either frames or series.

`FromTable` — fits quite well as a SQL-ism, maybe grammatically a bit odd in the context of the other operations

`UseTable` — avoids the confusing presence of "select" but maybe not much clearer

I feel like `SelectColumn` doesn't suffer from the same confusion, but let me know if you disagree.

Reply:
I too like fact that frame and series don't appear.

The symmetry between `SelectTable` and `SelectColumn` seems helpful. But maybe the collision with SQL's `Select` is unhelpful.

I also like the fact that all the nodes represent operations and are named for verbs (except the functions which I _suppose_ could be renamed to `ApplyX`).

As well as "use", which Dave suggests, I suppose we could try "pick" or
 "choose".

Reply:
I'm -0 on `SelectTable` and +0 on `UseTable` or `GetTable`.



> I feel like `SelectColumn` doesn't suffer from the same confusion, but let me know if you disagree.

Agree.